hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -100,6 +141,338 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
+	// whatever reason, we will not know until the next update, at which point hopefully the MCD is able
+	// to reconcile
+	for _, node := range nodes {
+		if node.Annotations[MachineConfigDaemonStateAnnotationKey] == MachineConfigDaemonStateDegraded {
+			setStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolCondition{
+				Type:               hyperv1.NodePoolUpdatingVersionConditionType,
+				Status:             corev1.ConditionFalse,
+				Reason:             hyperv1.NodePoolInplaceUpgradeFailedConditionReason,
+				Message:            fmt.Sprintf("Node %s in nodepool degraded: %v", node.Name, node.Annotations[MachineConfigDaemonReasonAnnotationKey]),
+				ObservedGeneration: nodePool.Generation,
+			})
+			return fmt.Errorf("degraded node found, cannot progress in-place upgrade")
+		}
+	}
+
+	// Create necessary update manifests, if they do not exist
+	err = r.createInPlaceUpdateManifests(ctx, hostedClusterClient, targetConfigVersionHash, ignEndpoint, caCertBytes, tokenBytes, nodePool)
+	if err != nil {
+		return fmt.Errorf("failed to create update manifests in hosted cluster: %v", err)
+	}
+
+	// Check the nodes to see if any need our help to progress drain
+	// TODO actually implement drain logic, likely as separate goroutines to monitor success
+	// TODO consider what happens if the desiredConfig has changed since the node last updated
+	for _, node := range nodes {
+		if node.Annotations[DesiredDrainerAnnotationKey] != node.Annotations[LastAppliedDrainerAnnotationKey] {
+			err = r.handleNodeDrainRequest(ctx, hostedClusterClient, node, node.Annotations[DesiredDrainerAnnotationKey])
+			if err != nil {
+				return fmt.Errorf("failed to create update manifests in hosted cluster: %v", err)
+			}
+			// TODO in the future, consider exiting here and let future syncs handle post-drain functions
+		}
+	}
+
+	// Find nodes that can be upgraded
+	// TODO add logic to honor maxUnavailable
+	nodesToUpgrade := getAllCandidateMachines(nodes, targetConfigVersionHash, 1)
+	err = r.setUpdateForNodes(ctx, hostedClusterClient, nodePool, nodesToUpgrade, targetConfigVersionHash)
+	if err != nil {
+		return fmt.Errorf("failed to set hosted nodes for inplace upgrade: %v", err)
+	}
+
+	return nil
+}
+
+func (r *NodePoolReconciler) setUpdateForNodes(ctx context.Context, hostedClusterClient client.Client, nodePool *hyperv1.NodePool, nodes []*corev1.Node, targetConfigVersionHash string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	for _, node := range nodes {
+		// Set the upgrade pod
+		// TODO maybe this can be a daemonset instead, since we are using a state machine MCD now
+		// There are also considerations on how to properly handle multiple updates, or to force updates
+		// on degraded nodes, etc.
+		namespace := inPlaceUpgradeNamespace(nodePool)
+		pod := inPlaceUpgradePod(namespace.Name, node.Name)
+		if result, err := controllerutil.CreateOrPatch(ctx, hostedClusterClient, pod, func() error {
+			return r.reconcileUpgradePod(
```

Comment:
where are we enforcing this pod runs only in the target node?

Reply:
> where are we enforcing this pod runs only in the target node?
The pod selects to run via
```
pod.Spec.NodeSelector = map[string]string{
		"kubernetes.io/hostname": nodeName,
	}
```
but we can run the pod on all nodes at all times. They won't update without explicit instruction from the controller, so we can make it run from a daemonset instead.

> which service account and perms are this pods using?
we are using the node's kubelet perms via `"--kubeconfig=/var/lib/kubelet/kubeconfig",` flag

Reply:
>pod.Spec.NodeSelector = map[string]string{
		"kubernetes.io/hostname": nodeName,
	}
>we are using the node's kubelet perms via "--kubeconfig=/var/lib/kubelet/kubeconfig", flag

Awesome, thanks.